### ISSUE TITLE
[MockProver] Allow advice columns on unusable rows to remain unassigned in instantiated gates

### ISSUE THREAD

**Issue by Brechtpd** [zcash/halo2#533]
These cells cannot be assigned to, and because they are marked as poisoned it also seems like there is no need to try and check that on a just per gate level with all cells values being marked as used even if that isn't actually the case. For other cells it is a reasonable assumption that all referenced cells should be assigned to (and a dummy value can be provided if needed) and if one would be forgotten it would not be obvious, but with these poisoned cells I don't think that can be the case.

https://github.com/zcash/halo2/blob/406f622e330e23ff91d645d43725e55de665c8e3/halo2_proofs/src/dev.rs#L768-L778

I think something like this would be fine:
```
// Check that it was assigned!
if r.cells.contains(&(cell.column, cell_row)) {
    None
} else if !self.usable_rows.contains(&cell_row) && matches!(cell.column.column_type(), Any::Advice) {
    None
} else {
    Some(VerifyFailure::CellNotAssigned {
        gate: (gate_index, gate.name()).into(),
        region: (r_i, r.name.clone()).into(),
        column: cell.column,
        offset: cell_row as isize - r.rows.unwrap().0 as isize,
    })
}
```

**Comment by therealyingtong** [comment]
I don't think we can hit a poisoned row in this code snippet, because it is within an iterator over `r.enabled_selectors` -- and if a selector had been enabled on a poisoned row, we would already have gotten an `Error::NotEnoughRowsAvailable`.

**Comment by daira** [comment]
I think @Brechtpd is probably referring to the case where a gate selected on a usable row, references a cell on an unusable row. But that is a circuit bug: it *should* be reported as an error. The way to avoid the error is to not select gates that overlap unusable rows.

**Comment by str4d** [comment]
This is definitely a case of "the `MockProver` is not smart enough yet", rather than a real restriction. In this case, it can't see that `q_enabled * (1 - q_first)` means that the constraint will never be active on the first row (where its prev query wraps around to query the unusable row).

**Comment by daira** [comment]
test: $E = m \cdot c^2$

**Comment by str4d** [comment]
While implementing #556 I remembered that I'd implemented `Constraints::with_selector` in a way that technically allows you to pass anything as a selector (as it just takes an `Expression<F>`). So it is now possible to write gate constraints in a way that separates the selector AST from the rest of the constraint, which we can and should leverage in `MockProver` for better analysis.
